Revamp of OSCompatibilityLayer filesystem functions.

Zemurin opened the pull request:
This is the first of possibly several *invasive* revamps to OSCompatiblityLayer.
- Functions that address filesystem have been unified, thanx to std::filesystem, with the exception of GetAllFilesInFolder as that one relies on native string/wstring names.
- function names have been partially standardized to capital first letter. This affects doesFileExist (thus invasive)
- some common functions dealing with files have been moved from eu4tovic2/ck2toeu4 as they're useful everywhere
- related tests have been moved

Git diff is largely useless due to clang changing everything.

Idhrendur commented on CommonFunctions.cpp:
```diff
@@ -0,0 +1,50 @@
+
+std::string replaceCharacter(std::string fileName, char character)
+{
+	auto position = fileName.find_first_of(character);
```
Update to instead use std::replace() from \<algorithm\>

Idhrendur commented on OSCompatibilityLayer.h:
```diff
@@ -47,67 +45,65 @@ typedef int errno_t;
+bool DeleteFolder(const std::string& folder);
+
+
+std::optional<std::string> GetFileFromTag(const std::string& directoryPath, const std::string& tag);
```
This seems to not really fit the overall theme here. As well, it seems unused in any of the current converters. Consider deleting it, or at the very least deprecating it.

Idhrendur commented on OSCompatibilityLayer.h:
```diff
@@ -1,37 +1,35 @@
+// used, the compiler will give an error due to sprintf_s_Linux returning void.
 #define sprintf_s sprintf_s_Linux
-void sprintf_s_Linux (char *__restrict __s, size_t __maxlen, const char *__restrict __format, ...);
+void sprintf_s_Linux(char* __restrict __s, size_t __maxlen, const char* __restrict __format, ...);
```
Double-check this, but I don't think sprintf is used anywhere any more. If not, go ahead and strip this out.

Zemurin replied:
Not yet. I'm doing all these changes from Window side at the moment, for a full linux cleanup I want to do it from a Linux environment. Only thing I changed so far on LinuxUtils was the common filesystem functions that I know will work on both OS-es. I can't even determine if this is used anywhere as the functions aren't directly referenced and the file isn't included on either of the converters.

Idhrendur commented on OSCompatibilityLayer.h:
```diff
@@ -1,37 +1,35 @@
 
-//Very basic implementation, simply returns 0 if FILE* is not NULL
+// Very basic implementation, simply returns 0 if FILE* is not NULL
 #define fopen_s fopen_s_Linux
```
Sadly, Vic2ToHoI4 is still using fopen_s for flags. I can look to fix that in a future cleanup.

Idhrendur commented on OSCompatibilityLayer.h:
```diff
@@ -47,67 +45,65 @@ typedef int errno_t;
+bool TryCreateFolder(const std::string& path);
+std::string GetCurrentDirectory();
+// Adds all files (just the file name) in the specified folder to the given collection.
+void GetAllFilesInFolder(const std::string& path, std::set<std::string>& fileNames);
```
While we're cleaning up this whole thing, replace this with
```std::set<std::string> GetAllFilesInFolder(const std::string& path)```

Keep the original (but marked deprecated) and make it work by calling the new one.

Idhrendur commented on OSCompatibilityLayer.h:
```diff
@@ -47,67 +45,65 @@ typedef int errno_t;
+std::string GetCurrentDirectory();
+// Adds all files (just the file name) in the specified folder to the given collection.
+void GetAllFilesInFolder(const std::string& path, std::set<std::string>& fileNames);
+void GetAllSubfolders(const std::string& path, std::set<std::string>& subfolders);
```
Ditto

Idhrendur commented on OSCompatibilityLayer.h:
```diff
@@ -47,67 +45,65 @@ typedef int errno_t;
+
+// For the specified folder and all subfolders, adds all files (just the subdirectory and file name)
+// to the given collection.
+void GetAllFilesInFolderRecursive(const std::string& path, std::set<std::string>& filenames);
```
Ditto

Idhrendur commented on LinuxUtils.cpp:
```diff
@@ -60,658 +60,587 @@ void fprintf_s_Linux(FILE* file, const char* format, ...)
+/*
+	Implemented all messages for error codes used in LinuxUtils.cpp.
+*/
+std::string GetLastErrorString()
```
Can maybe use [std::strerror()](https://en.cppreference.com/w/cpp/string/byte/strerror) instead. Though I thought I saw something else recently.

Zemurin replied:
strerror_s is the secure one, that one is depreciated. I've changed the function but haven't tested it yet. Need linux environment for that,

Idhrendur commented on OSCompatibilityLayer.h:
```diff
@@ -1,37 +1,35 @@
 #ifndef OS_COMPATABILITY_LAYER_H
```
Tests aren't strictly necessary, as they haven't existed before, but I wouldn't object if you added them.

I've recently added a few tests to Vic2ToHoI4 that do filesystem things and that can serve as a guide. You'll need to include test data and be aware of where it gets placed and where the tests are running, but it can be done. Visual Studio is actually pretty good about placing files if you include them in the project. I've been doing that in Vic2ToHoI4 for the Windows build for everything but blankMod. Sadly I don't know how to do it in CMake for the Linux build.

Zemurin replied:
Yeah, I'm not so hot about doing OS-level tests. 

These functions, as they are now, in OSCommonLayer.cpp, they are wrappers around std::filesystem, with added logging, sometimes not even that (DoesFileExist made sense before, but now it's just a fs::exists()). 

Also they are used commonly throughout all converters so any deviance would immediately be spotted as conversions would fail.

If there's a specific function relating to a specific task, like renaming something to something else, depending on user input, then sure, we can test that in the scope of that converter function, but for global OS-level functions, testing them seems redundant - they either work or they don't and if they don't converter won't run at all, ever.

Zemurin commented:
I've pushed all changes in and would be happy if you took another peek. 

Ignore LinuxUtils for now, I need to establish a linux build environment before I can clean it out properly. So far the only (I hope) invasive change is renaming capitalizations for functions, bit that's just a matter of search/replace when integrating new commonItems.

As for GetFileFromTag, it was still used in eu4tovic2, under Country.cpp, but I moved the function over and cleaned this up. It was an eyesore for me as well, but I didn't know if you used it in hoi4, so was hesitant to delete it.

Idhrendur commented on CardinalToOrdinal.h:
```diff
@@ -1,14 +0,0 @@
-#ifndef CARDINAL_TO_ORDINAL_H
```
Need the ```#include "CardinalToOrdinal.h"``` and pragma warning here too.
